tsl/src/continuous_aggs/refresh.c

```diff
@@ -326,8 +383,39 @@ continuous_agg_refresh_with_window(const ContinuousAgg *cagg,
 		InternalTimeRange bucketed_refresh_window =
 			compute_circumscribed_bucketed_refresh_window(&invalidation, cagg->data.bucket_width);
 
-		log_refresh_window(DEBUG1, cagg, &bucketed_refresh_window, "invalidation refresh on");
-		continuous_agg_refresh_execute(&refresh, &bucketed_refresh_window, chunk_id);
+		if (do_merged_refresh)
+		{
+			if (count == 0)
+				merged_refresh_window = bucketed_refresh_window;
+			else
+			{
+				if (bucketed_refresh_window.start < merged_refresh_window.start)
+					merged_refresh_window.start = bucketed_refresh_window.start;
+
+				if (bucketed_refresh_window.end > merged_refresh_window.end)
+					merged_refresh_window.end = bucketed_refresh_window.end;
+			}
+		}
+		else
+		{
+			log_refresh_window(DEBUG1, cagg, &bucketed_refresh_window, "invalidation refresh on");
+			continuous_agg_refresh_execute(&refresh, &bucketed_refresh_window, chunk_id);
+		}
+
+		count++;
+	}
+
+	if (do_merged_refresh && count > 0)
+	{
+		Assert(merged_refresh_window.type == refresh_window->type);
+		Assert(merged_refresh_window.start >= refresh_window->start);
+		Assert(merged_refresh_window.end <= refresh_window->end);
+
+		log_refresh_window(DEBUG1,
+						   cagg,
+						   &merged_refresh_window,
+						   psprintf("merged %ld invalidations for refresh on", count));
```

Comment:
I think this should be a notice (https://www.postgresql.org/docs/12/runtime-config-logging.html#RUNTIME-CONFIG-SEVERITY-LEVELS):

> NOTICE | Provides information that might be helpful to users, e.g., notice of truncation of long identifiers.

Reply:
I disagree. A notice is sent to the client and this is not normally valuable information to the user since it is implementation detail (debug information). Users care that the result of the refresh is what they expect, not _how_ it was executed. If they want to see this info, they can always change the debug level.

Reply:
The example given in the manual is where something happened that might change the behavior but not likely affect the result, e.g., truncation of identifier. The PostgreSQL code uses notices internally for these things (and others, but these are examples):

- Operations are skipped for statements, e.g., table is not dropped in a drop statement
- Constraints are merged with inherited definition
- Merging multiple inherited definitions

I think this message belongs in the same category.

Reply:
This is about how we internally process invalidations. Users might not even know what an invalidation is, and most won't care.